### PR TITLE
Have DOLT_PATCH() system table function implement an index on data change vs schema change.

### PR DESCRIPTION
**Pull request by nicktobey**
This allows for efficient queries of the form:

`SELECT * FROM DOLT_PATCH(...) WHERE diff_type = "schema"`

Dolt will only compute the schema changes and avoid computing the data changes, which may be expensive.

### CODE REVIEW

**Review by max-hoffman — APPROVED** [review]
I think the changes look good. The last thing I'd do on this side is: 1) make sure the GMS PR update for this doesn't break any of our integration and compatibility tests, and 2) make a test suite that tests for indexed table functions (and correctness? not seeing any correctness tests right now), something like `TestIndexedTableFunctions`, and transfer the current patch tests. That should make it easier for others to add and test table function indexes in a thorough way in the future.

File: go/libraries/doltcore/sqle/dolt_patch_table_function.go [hunk 1]
```diff
@@ -50,13 +71,143 @@ type PatchTableFunction struct {
...
+// LookupPartitions is a sql.IndexedTable interface function that takes an index lookup and returns the set of corresponding partitions.
+func (p *PatchTableFunction) LookupPartitions(context *sql.Context, lookup sql.IndexLookup) (sql.PartitionIter, error) {
+	if lookup.Index.ID() == diffTypeColumnName {
+		diffTypes, ok := index.LookupToPointSelectStr(lookup)
+		if !ok {
+			return nil, fmt.Errorf("failed to parse commit lookup ranges: %s", sql.DebugString(lookup.Ranges))
+		}
+
+		includeSchemaDiff := slices.Contains(diffTypes, "schema")
+		includeDataDiff := slices.Contains(diffTypes, "data")
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/libraries/doltcore/sqle/dolt_patch_table_function.go:166]: No. I went and created some.

File: go/libraries/doltcore/sqle/enginetest/dolt_queries_diff.go [hunk 2]
```diff
@@ -2963,6 +2963,66 @@ var PatchTableFunctionScriptTests = []queries.ScriptTest{
...
+					{" │       └─ columns: [name t]"},
+					{" └─ TableAlias(dolt_patch)"},
+					{"     └─ IndexedTableAccess(DOLT_PATCH(@Commit0, @Commit0, 't'))"},
+					{"         └─ index: [DOLT_PATCH(@Commit0, @Commit0, 't').diff_type]"},
+				},
+			},
+			{
+				Query: "describe SELECT ( SELECT statement FROM (SELECT * FROM dolt_patch(@Commit0, @Commit0, 't') where diff_type = diff_type_name.t) as rhs) from diff_type_name;",
+				Expected: []sql.Row{
+					{"Project"},
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/libraries/doltcore/sqle/enginetest/dolt_queries_diff.go]: Okay, replaced the tests with better ones, and moved them into their own test suite for indexed system table functions.

**Comment by nicktobey** [comment]
Fixes https://github.com/dolthub/dolt/issues/6446

**Pull request merged**